### PR TITLE
EN-13280: added endpoint for bulk accounts request

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- Previously, if wanting to check the state of N addresses, you would have had to perform N requests.
  
## Proposed Changes
- Added an endpoint that can return the state of a bulk of addresses

## Testing procedure
- perform a query similar to:
```
curl --request POST \
  --url http://127.0.0.1:8070/address/bulk \
  --header 'Content-Type: application/json' \
  --data '[
	"erd1e2hdx0y6hk5nenw6r8f2msatnkd2gzdl8e77jd42f30rqndwks5qcp9sgg",
	"erd1z5dvmhj4l56sdwym2f8sdej6d4hnfk6r9ylcvnsa3eeaznr4fvlqxevnyv",
	"erd1aej77kzx2spz0trnfkc0m52zr3yw4tw48hkf36nmhzvv9kgtyyuqdeh6es",
	"erd15t7d6nu2753kc34505qrsv80gp5yuz5j8hr7eg25fh7w4vjezgjqy4cj48",
	"erd1sxnlcwakg6ah6q7r8zv6z920u8282emvpt40daezf2z6vtn3adsq9f09u4"
]'
```

### CODE REVIEW

File: api/groups/addressGroup.go [hunk 1]
```diff
@@ -180,6 +187,30 @@ func (ag *addressGroup) getAccount(c *gin.Context) {
 	shared.RespondWithSuccess(c, gin.H{"account": accountResponse, "blockInfo": blockInfo})
 }
 
+// getAccounts returns the state of the provided addresses on the specified block
+func (ag *addressGroup) getAccounts(c *gin.Context) {
```
Comment (andreibancioiu) [inline api/groups/addressGroup.go:191]: :+1:

File: facade/nodeFacade.go [hunk 2]
```diff
@@ -337,6 +337,41 @@ func (nf *nodeFacade) GetAccount(address string, options apiData.AccountQueryOpt
 	return accountResponse, blockInfo, nil
 }
 
+// TODO: refactor GetAccounts. Make AccountRepository/AccountsAdapterAPI able to return the entire accounts adapter in order to
+// only fetch it once, and then re-use it for getting all the accounts.
+// Currently, there is no improvement in changing the existing mechanism (GetAccount -> GetAccount*s*) since it would
+// still mean the recreation of the trie for each of the account
```
Comment (andreibancioiu) [inline facade/nodeFacade.go:343]: It should be _somehow_ all right, though since the `latest recreated trie` is cached.

Comment (andreibancioiu) [reply facade/nodeFacade.go:343]: Plus, there aren't `N` HTTP requests anymore, which is a nice improvement by itself.

File: facade/nodeFacade.go [hunk 3]
```diff
@@ -337,6 +337,41 @@ func (nf *nodeFacade) GetAccount(address string, options apiData.AccountQueryOpt
...
+
+// GetAccounts returns the state of the provided addresses
+func (nf *nodeFacade) GetAccounts(addresses []string, options apiData.AccountQueryOptions) (map[string]*apiData.AccountResponse, apiData.BlockInfo, error) {
+	numAddresses := uint32(len(addresses))
+	// TODO: check if Antiflood is enabled before applying this constraint (EN-13278)
```
Comment (andreibancioiu) [inline facade/nodeFacade.go:348]: :+1:

File: facade/nodeFacade.go [hunk 4]
```diff
@@ -337,6 +337,41 @@ func (nf *nodeFacade) GetAccount(address string, options apiData.AccountQueryOpt
...
+	if numAddresses > maxBulkSize {
+		return nil, apiData.BlockInfo{}, fmt.Errorf("%w (provided: %d, maximum: %d)", ErrTooManyAddressesInBulk, numAddresses, maxBulkSize)
+	}
+
+	response := make(map[string]*apiData.AccountResponse)
+	var blockInfo apiData.BlockInfo
+
+	for _, address := range addresses {
+		accountResponse, blockInfoForAccount, err := nf.node.GetAccount(address, options)
+		if err != nil {
```
Comment (andreibancioiu) [inline facade/nodeFacade.go:359]: This is good. If one fetch fails, the whole call should fail :+1:

File: facade/nodeFacade.go [hunk 5]
```diff
@@ -337,6 +337,41 @@ func (nf *nodeFacade) GetAccount(address string, options apiData.AccountQueryOpt
...
+			return nil, apiData.BlockInfo{}, err
+		}
+
+		blockInfo = blockInfoForAccount
+
+		codeHash := accountResponse.CodeHash
+		code, _ := nf.node.GetCode(codeHash, options)
```
Comment (andreibancioiu) [inline facade/nodeFacade.go:366]: Oh, se we are including the `code`, as well.

**Review by andreibancioiu — COMMENTED** [review]
Perhaps adjust the TODO, otherwise :christmas_tree: 

:+1:

**Review by gabi-vuls — APPROVED** [review]
System test passed.

File: facade/nodeFacade.go (hunk 2 above)
Comment (bogdan-rosianu) [reply facade/nodeFacade.go]: removed todo

**Review by ssd04 — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**